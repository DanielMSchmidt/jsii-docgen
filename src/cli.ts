import * as fs from 'fs';
import * as yargs from 'yargs';
import { Language } from './docgen/transpile/transpile';
import { Documentation } from './index';

type GenerateOptions = {
  readme?: boolean;
  language: Language;
  submodule?: string;
  allSubmodules?: boolean;
}

async function generateForLanguage(docs: Documentation, format: 'md' | 'json', options: GenerateOptions, output = 'API') {
  const content = await (format === 'md' ? docs.toMarkdown(options) : docs.toJson(options));
  const fileSuffix = format === 'md' ? 'md' : 'json';

  fs.writeFileSync(`${output ?? 'API'}.${fileSuffix}`, content.render());
}


export async function main() {
  const args = yargs
    .usage('Usage: $0')
    .option('output', { alias: 'o', type: 'string', required: false, desc: 'Output filename, the file type is automatically added. (defaults to API.md if format is markdown, and API.json if format is JSON and one language is passed, API.language.md and API.language.json if multilpe are passed)' })
    .option('format', { alias: 'f', default: 'md', choices: ['md', 'json'], desc: 'Output format, markdown or json' })
    .option('language', { array: true, alias: 'l', default: ['typescript'], choices: Language.values().map(x => x.toString()), desc: 'Output language' })
    .option('package', { alias: 'p', type: 'string', required: false, desc: 'The name@version of an NPM package to document', defaultDescription: 'The package in the current directory' })
    .option('submodule', { alias: 's', type: 'string', required: false, desc: 'Generate docs for a specific submodule (or "root")' })
    .example('$0', 'Generate documentation for the current module as a single file (auto-resolves node depedencies)')
    .argv;


  const submodule = args.submodule === 'root' ? undefined : args.submodule;
  const allSubmodules = !args.submodule;
  const docs = await (args.package
    ? Documentation.forPackage(args.package)
    : Documentation.forProject(process.cwd()));
  const options = (lang: string) => ({ readme: false, language: Language.fromString(lang), submodule, allSubmodules });

  if (args.language.length <= 1) {
    await generateForLanguage(docs, args.format as 'md' | 'json', options(args.language[0]), args.output);
  } else {
    for (const lang of args.language) {
      const output = `${args.output ?? 'API'}.${lang}`;
      await generateForLanguage(docs, args.format as 'md' | 'json', options(lang), output);
    }
  }
}

main().catch(e => {
  console.error(e);
  process.exit(1);
});
