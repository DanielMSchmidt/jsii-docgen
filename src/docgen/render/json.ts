/**
 * Type-safe Json renderer.
 */
export class Json<T> {
  constructor(private readonly content: T) {};

  public render(): string {
    return JSON.stringify(this.content, null, 2);
  }
}