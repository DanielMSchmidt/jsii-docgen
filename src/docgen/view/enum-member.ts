import * as reflect from 'jsii-reflect';
import { EnumMemberSchema, extractDocs } from '../schema';
import { Transpile, TranspiledEnumMember } from '../transpile/transpile';

export class EnumMember {
  private readonly transpiled: TranspiledEnumMember;
  constructor(transpile: Transpile, private readonly em: reflect.EnumMember) {
    this.transpiled = transpile.enumMember(em);
  }

  public toJson(): EnumMemberSchema {
    return {
      jsiiId: `${this.em.enumType.fqn}.${this.em.name}`,
      displayName: this.transpiled.name,
      fqn: this.transpiled.fqn,
      docs: extractDocs(this.em.docs),
      submoduleJsiiId: this.transpiled.type.submoduleJsiiId,
      submoduleFqn: this.transpiled.type.submoduleFqn,
      sourceFile: this.transpiled.type.source.locationInModule?.filename,
      sourceLine: this.transpiled.type.source.locationInModule?.line,
    };
  }
}
