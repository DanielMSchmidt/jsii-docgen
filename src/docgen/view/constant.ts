import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { PropertySchema } from '../schema';
import { Transpile, TranspiledType } from '../transpile/transpile';
import { Property } from './property';

export class Constant {
  private readonly constant: Property;
  constructor(transpile: Transpile, property: reflect.Property, linkFormatter: (type: TranspiledType) => string) {
    this.constant = new Property(transpile, property, linkFormatter);
  }
  public toMarkdown(): Markdown {
    return this.constant.toMarkdown();
  }

  public toJson(): PropertySchema {
    return this.constant.toJson();
  }
}
