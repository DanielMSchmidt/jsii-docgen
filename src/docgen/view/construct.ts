import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { ConstructSchema } from '../schema';
import { Transpile, TranspiledType } from '../transpile/transpile';
import { Class } from './class';

export class Construct {
  private readonly construct: Class;
  constructor(transpile: Transpile, klass: reflect.ClassType, linkFormatter: (type: TranspiledType) => string) {
    this.construct = new Class(transpile, klass, linkFormatter);
  }
  public toMarkdown(): Markdown {
    return this.construct.toMarkdown();
  }

  public toJson(): ConstructSchema {
    return this.construct.toJson();
  }
}
