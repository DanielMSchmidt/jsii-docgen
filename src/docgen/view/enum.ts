import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { EnumSchema } from '../schema';
import { Transpile } from '../transpile/transpile';
import { EnumMember } from './enum-member';

export class Enum {
  private readonly members: EnumMember[];
  constructor(
    private readonly transpile: Transpile,
    private readonly enu: reflect.EnumType,
  ) {
    this.members = enu.members.map((em) => new EnumMember(transpile, em));
  }

  public toMarkdown(): Markdown {
    const transpiled = this.transpile.enum(this.enu);
    const md = new Markdown({ header: { title: transpiled.name } });

    if (this.enu.docs) {
      md.docs(this.enu.docs);
    }

    for (const m of this.members) {
      md.section(m.toMarkdown());
    }

    return md;
  }

  public toJson(): EnumSchema {
    const transpiled = this.transpile.enum(this.enu);
    return {
      name: transpiled.name,
      docs: this.enu.docs.toString(),
      members: this.members.map((member) => member.toJson()),
    };
  }
}
