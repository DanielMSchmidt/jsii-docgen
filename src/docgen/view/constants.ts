import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { PropertySchema } from '../schema';
import { Transpile, TranspiledType } from '../transpile/transpile';
import { Constant } from './constant';

export class Constants {
  private readonly constants: Constant[];
  constructor(transpile: Transpile, properties: reflect.Property[]) {
    this.constants = properties
      .filter((p) => !p.protected && p.const)
      .map((p) => new Constant(transpile, p));
  }

  public toMarkdown(linkFormatter: (type: TranspiledType) => string): Markdown {
    if (this.constants.length === 0) {
      return Markdown.EMPTY;
    }

    const md = new Markdown({ header: { title: 'Constants' } });
    for (const c of this.constants) {
      md.section(c.toMarkdown(linkFormatter));
    }
    return md;
  }

  public toJson(): PropertySchema[] {
    return this.constants.map((constant) => constant.toJson());
  }
}
