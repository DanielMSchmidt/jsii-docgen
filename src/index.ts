// tslint:disable:no-console
import * as path from 'path';
import * as fs from 'fs-extra';
import * as jsiiReflect from 'jsii-reflect';
import { EnumPage } from './render/enum';
import { Home } from './render/home';
import { ClassPage, InterfacePage } from './render/klass';
import { elementAnchorLink } from './render/links';
import { Page, RenderContext, JsiiEntity } from './render/page';
import { flatMap } from './render/util';

/**
 * Renders markdown files into an output directory for a jsii typesystem.
 * @param jsiiFiles jsii input files
 * @param outdir output directory
 */
export async function renderFiles(jsiiFiles: string[], outdir: string) {
  const renderFileName = (type: JsiiEntity) => type.fqn.replace('/', '_') + '.md';

  const ctx: RenderContext = {
    links: {
      renderLink: type => `./${renderFileName(type)}`,
    },
  };

  const ts = new jsiiReflect.TypeSystem();
  for (const filePath of jsiiFiles) {
    await ts.load(filePath);
  }

  const pages = await renderPages(ts, ctx);
  for (const page of pages) {
    const filePath = path.join(outdir, renderFileName(page.type));
    console.log(filePath);
    await fs.mkdirp(path.dirname(filePath));
    await fs.writeFile(filePath, page.markdown, { encoding: 'utf-8' });
  }
}

export async function renderSinglePageModule(moduleDir: string, outFile: string) {
  const ts = new jsiiReflect.TypeSystem();
  const asm = await ts.loadModule(moduleDir);

  const ctx: RenderContext = {
    readme: false,
    heading: 2,
    links: {
      renderLink: type => {
        if (type instanceof jsiiReflect.Assembly) {
          return '';
        } else {
          return elementAnchorLink(type);
        }
      },
    },
  };

  const file = fs.createWriteStream(outFile);

  const pages = await renderPages(ts, ctx);
  for (const page of pages) {
    if (page.type.fqn.split('.')[0] !== asm.fqn) {
      continue; // skip other modules
    }

    file.write(page.markdown);
    file.write('\n\n\n');
  }

  file.close();
}

/**
 * Renders markdown reference documentation pages for all the types in this jsii
 * typesystem.
 * @param typesystem the jsii typesystem to reflect on.
 * @param ctx render context
 */
export async function renderPages(typesystem: jsiiReflect.TypeSystem, ctx: RenderContext): Promise<Page[]> {
  const result = new Array<Page>();

  for (const assembly of typesystem.assemblies) {
    result.push(...documentAssembly(ctx, assembly));
  }

  return result;
}

function documentAssembly(ctx: RenderContext, asm: jsiiReflect.Assembly): Page[] {
  const classes = [...asm.classes, ...flatMap(asm.submodules, submod => [...submod.classes])];
  const interfaces = [...asm.interfaces, ...flatMap(asm.submodules, submod => [...submod.interfaces])];
  const enums = [...asm.enums, ...flatMap(asm.submodules, submod => [...submod.enums])];

  return [
    new Home(ctx, asm),
    ...classes.map(c => new ClassPage(ctx, c)),
    ...interfaces.map(i => new InterfacePage(ctx, i)),
    ...enums.map(e => new EnumPage(ctx, e)),
  ];
}

