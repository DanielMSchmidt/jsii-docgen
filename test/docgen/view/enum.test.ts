import * as reflect from 'jsii-reflect';
import { AssemblyMetadataSchema } from '../../../src';
import { MarkdownRenderer } from '../../../src/docgen/render/markdown-render';
import { Language } from '../../../src/docgen/transpile/transpile';
import { LANGUAGE_SPECIFIC } from '../../../src/docgen/view/documentation';
import { Enum } from '../../../src/docgen/view/enum';
import { Assemblies } from '../assemblies';

const assembly: reflect.Assembly = Assemblies.instance.withoutSubmodules;

const metadata: AssemblyMetadataSchema = {
  jsiiProjectName: assembly.name,
  jsiiProjectVersion: assembly.version,
  jsiiProjectLanguages: Object.keys(assembly.targets ?? []),
  packageName: assembly.targets?.python?.distName,
  packageVersion: assembly.version,
  moduleFqn: assembly.targets?.python?.module,
};

const findEnum = (): reflect.EnumType => {
  if (assembly.enums[0]) {
    return assembly.enums[0];
  }
  throw new Error('Assembly does not contain an emum');
};

test.each(Language.values())('%s snapshot', (language) => {
  const { transpile } = LANGUAGE_SPECIFIC[language.toString()];
  const markdown = new MarkdownRenderer({ language, ...metadata });
  const enu = new Enum(transpile, findEnum()).toJson();
  expect(enu).toMatchSnapshot();
  expect(markdown.visitEnum(enu).render()).toMatchSnapshot();
});
