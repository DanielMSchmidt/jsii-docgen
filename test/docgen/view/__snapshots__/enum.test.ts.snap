// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csharp snapshot 1`] = `
Object {
  "displayName": "TagMutability",
  "docs": Object {
    "summary": "The tag mutability setting for your repository.",
  },
  "fqn": "Amazon.CDK.AWS.ECR.TagMutability",
  "jsiiId": "@aws-cdk/aws-ecr.TagMutability",
  "members": Array [
    Object {
      "displayName": "MUTABLE",
      "docs": Object {
        "summary": "allow image tags to be overwritten.",
      },
      "fqn": "Amazon.CDK.AWS.ECR.TagMutability.MUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
    Object {
      "displayName": "IMMUTABLE",
      "docs": Object {
        "summary": "all image tags within the repository will be immutable which will prevent them from being overwritten.",
      },
      "fqn": "Amazon.CDK.AWS.ECR.TagMutability.IMMUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
  ],
  "sourceFile": "lib/repository.ts",
  "sourceLine": 653,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
}
`;

exports[`csharp snapshot 2`] = `
" TagMutability <a name=\\"TagMutability\\" id=\\"@aws-cdk/aws-ecr.TagMutability\\"></a>

The tag mutability setting for your repository.

# Members <a name=\\"Members\\" id=\\"Members\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>Amazon.CDK.AWS.ECR.TagMutability.MUTABLE</code> | allow image tags to be overwritten. |
| <code>Amazon.CDK.AWS.ECR.TagMutability.IMMUTABLE</code> | all image tags within the repository will be immutable which will prevent them from being overwritten. |

---

## \`MUTABLE\` <a name=\\"MUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.MUTABLE\\"></a>

allow image tags to be overwritten.

---


## \`IMMUTABLE\` <a name=\\"IMMUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\"></a>

all image tags within the repository will be immutable which will prevent them from being overwritten.

---

"
`;

exports[`go snapshot 1`] = `
Object {
  "displayName": "TagMutability",
  "docs": Object {
    "summary": "The tag mutability setting for your repository.",
  },
  "fqn": "github.com/aws/aws-cdk-go/awscdkawsecr.TagMutability",
  "jsiiId": "@aws-cdk/aws-ecr.TagMutability",
  "members": Array [
    Object {
      "displayName": "TagMutability_MUTABLE",
      "docs": Object {
        "summary": "allow image tags to be overwritten.",
      },
      "fqn": "github.com/aws/aws-cdk-go/awscdkawsecr.TagMutability_MUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
    Object {
      "displayName": "TagMutability_IMMUTABLE",
      "docs": Object {
        "summary": "all image tags within the repository will be immutable which will prevent them from being overwritten.",
      },
      "fqn": "github.com/aws/aws-cdk-go/awscdkawsecr.TagMutability_IMMUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
  ],
  "sourceFile": "lib/repository.ts",
  "sourceLine": 653,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
}
`;

exports[`go snapshot 2`] = `
" TagMutability <a name=\\"TagMutability\\" id=\\"@aws-cdk/aws-ecr.TagMutability\\"></a>

The tag mutability setting for your repository.

# Members <a name=\\"Members\\" id=\\"Members\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>github.com/aws/aws-cdk-go/awscdkawsecr.TagMutability_MUTABLE</code> | allow image tags to be overwritten. |
| <code>github.com/aws/aws-cdk-go/awscdkawsecr.TagMutability_IMMUTABLE</code> | all image tags within the repository will be immutable which will prevent them from being overwritten. |

---

## \`TagMutability_MUTABLE\` <a name=\\"TagMutability_MUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.MUTABLE\\"></a>

allow image tags to be overwritten.

---


## \`TagMutability_IMMUTABLE\` <a name=\\"TagMutability_IMMUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\"></a>

all image tags within the repository will be immutable which will prevent them from being overwritten.

---

"
`;

exports[`java snapshot 1`] = `
Object {
  "displayName": "TagMutability",
  "docs": Object {
    "summary": "The tag mutability setting for your repository.",
  },
  "fqn": "software.amazon.awscdk.services.ecr.TagMutability",
  "jsiiId": "@aws-cdk/aws-ecr.TagMutability",
  "members": Array [
    Object {
      "displayName": "MUTABLE",
      "docs": Object {
        "summary": "allow image tags to be overwritten.",
      },
      "fqn": "software.amazon.awscdk.services.ecr.TagMutability.MUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
    Object {
      "displayName": "IMMUTABLE",
      "docs": Object {
        "summary": "all image tags within the repository will be immutable which will prevent them from being overwritten.",
      },
      "fqn": "software.amazon.awscdk.services.ecr.TagMutability.IMMUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
  ],
  "sourceFile": "lib/repository.ts",
  "sourceLine": 653,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
}
`;

exports[`java snapshot 2`] = `
" TagMutability <a name=\\"TagMutability\\" id=\\"@aws-cdk/aws-ecr.TagMutability\\"></a>

The tag mutability setting for your repository.

# Members <a name=\\"Members\\" id=\\"Members\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>software.amazon.awscdk.services.ecr.TagMutability.MUTABLE</code> | allow image tags to be overwritten. |
| <code>software.amazon.awscdk.services.ecr.TagMutability.IMMUTABLE</code> | all image tags within the repository will be immutable which will prevent them from being overwritten. |

---

## \`MUTABLE\` <a name=\\"MUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.MUTABLE\\"></a>

allow image tags to be overwritten.

---


## \`IMMUTABLE\` <a name=\\"IMMUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\"></a>

all image tags within the repository will be immutable which will prevent them from being overwritten.

---

"
`;

exports[`python snapshot 1`] = `
Object {
  "displayName": "TagMutability",
  "docs": Object {
    "summary": "The tag mutability setting for your repository.",
  },
  "fqn": "aws_cdk.aws_ecr.TagMutability",
  "jsiiId": "@aws-cdk/aws-ecr.TagMutability",
  "members": Array [
    Object {
      "displayName": "MUTABLE",
      "docs": Object {
        "summary": "allow image tags to be overwritten.",
      },
      "fqn": "aws_cdk.aws_ecr.TagMutability.MUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
    Object {
      "displayName": "IMMUTABLE",
      "docs": Object {
        "summary": "all image tags within the repository will be immutable which will prevent them from being overwritten.",
      },
      "fqn": "aws_cdk.aws_ecr.TagMutability.IMMUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
  ],
  "sourceFile": "lib/repository.ts",
  "sourceLine": 653,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
}
`;

exports[`python snapshot 2`] = `
" TagMutability <a name=\\"TagMutability\\" id=\\"@aws-cdk/aws-ecr.TagMutability\\"></a>

The tag mutability setting for your repository.

# Members <a name=\\"Members\\" id=\\"Members\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>aws_cdk.aws_ecr.TagMutability.MUTABLE</code> | allow image tags to be overwritten. |
| <code>aws_cdk.aws_ecr.TagMutability.IMMUTABLE</code> | all image tags within the repository will be immutable which will prevent them from being overwritten. |

---

## \`MUTABLE\` <a name=\\"MUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.MUTABLE\\"></a>

allow image tags to be overwritten.

---


## \`IMMUTABLE\` <a name=\\"IMMUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\"></a>

all image tags within the repository will be immutable which will prevent them from being overwritten.

---

"
`;

exports[`typescript snapshot 1`] = `
Object {
  "displayName": "TagMutability",
  "docs": Object {
    "summary": "The tag mutability setting for your repository.",
  },
  "fqn": "@aws-cdk/aws-ecr.TagMutability",
  "jsiiId": "@aws-cdk/aws-ecr.TagMutability",
  "members": Array [
    Object {
      "displayName": "MUTABLE",
      "docs": Object {
        "summary": "allow image tags to be overwritten.",
      },
      "fqn": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
    Object {
      "displayName": "IMMUTABLE",
      "docs": Object {
        "summary": "all image tags within the repository will be immutable which will prevent them from being overwritten.",
      },
      "fqn": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
      "jsiiId": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
      "sourceFile": "lib/repository.ts",
      "sourceLine": 653,
      "submoduleFqn": undefined,
      "submoduleJsiiId": undefined,
    },
  ],
  "sourceFile": "lib/repository.ts",
  "sourceLine": 653,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
}
`;

exports[`typescript snapshot 2`] = `
" TagMutability <a name=\\"TagMutability\\" id=\\"@aws-cdk/aws-ecr.TagMutability\\"></a>

The tag mutability setting for your repository.

# Members <a name=\\"Members\\" id=\\"Members\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>@aws-cdk/aws-ecr.TagMutability.MUTABLE</code> | allow image tags to be overwritten. |
| <code>@aws-cdk/aws-ecr.TagMutability.IMMUTABLE</code> | all image tags within the repository will be immutable which will prevent them from being overwritten. |

---

## \`MUTABLE\` <a name=\\"MUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.MUTABLE\\"></a>

allow image tags to be overwritten.

---


## \`IMMUTABLE\` <a name=\\"IMMUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\"></a>

all image tags within the repository will be immutable which will prevent them from being overwritten.

---

"
`;
